sympy/polys/ring_series.py

```diff
@@ -1158,3 +1158,97 @@ def rs_compose_add(p1, p2):
     if dp:
         q = q*x**dp
     return q
+
+
+class RingSeriesBase(object):
+    pass
+
+class RingSeries(RingSeriesBase):
+    """Sparse Multivariate Ring Series"""
+
+    def new(self, *args, **kwargs):
+        return self.__class__(*args, **kwargs)
+
+    def __init__(cls, series, degree=None):
+        cls.series = series
+        cls.ring = series.ring
+        if degree:
+            cls.degree = degree
+
+    def __repr__(self):
+        return "%s(%s, %s, %s)" % (self.__class__.__name__, self.series,
+            self.ring.domain, self.ring.gens)
+
+    def __pos__(self):
+        return self
+
+    def __neg__(self):
+        return self.new(-self.series, self.ring)
+
+    def __add__(self, other):
+        if self.ring == other.ring:
+            return self.new(self.series + other.series, self.ring)
+        else:
+            raise DomainError("Both the series should be defined on the same"
+                "ring")
+
+    def __mul__(self, other):
+        if self.ring == other.ring:
+            return self.new(self.series * other.series, self.ring)
+        else:
+            raise DomainError("Both the series should be defined on the same"
+                "ring")
+
+    def __sub__(self, other):
+        return self + (-other)
+
+    def get_degree(self):
+        if self.degree:
+            return self.degree
+        else:
+            self.degree = self.series.degree()
+            return self.degree
+
+class rs_sin(RingSeriesBase):
```

Comment:
@leosartaj I guess we can put some of the functions in `RingSeriesBase` to avoid duplication. `rs_sin` is just a symbolic rep of `sin`. `RingSeries` is what you get after you expand it. Do you think `rs_sin` should be independent of `RingSeriesBase` hierarchy?